### PR TITLE
Log batch errors in a specific file then requeue all

### PR DESCRIPTION
**Pull request by thomas-hilaire**

### CODE REVIEW

File: src/importer.ts [hunk 1]
```diff
@@ -30,35 +30,45 @@ export class Importer {
...
-            });
+        logger.info("Batch %d has %d events", runCount, amqpMessages.length);
+        this.runBatchOnPapi(amqpMessages).then(batchResult => {
+            logger.info("Batch %d is done: ", runCount, batchResult.message);
+            batchResult.errors.forEach(e => batchErrorLogger.error("Batch #%d", runCount, e));
+            setTimeout(() => this.ackOrRequeueAllIfAnyError(batchResult, amqpChannel, amqpMessages), this.config.delayBetweenBatchMs);
+        });
     }
 
+    private ackOrRequeueAllIfAnyError(batchResult, amqpChannel, amqpMessages) {
```
Comment (fvn-linagora) [inline src/importer.ts]: I suggest to not use RabbitMQ-related types/objects in message consumer implementation, and stick with business logic only (consumers/publishers should not always know about the underlying message broker IMO).
For instance, an abstraction can be passed that would handle delivering an error message to the configured queue, through a publish(Error) method.

Comment (thomas-hilaire) [inline src/importer.ts]: I do that in the next review so I don't want to refactor this one.
Note that I use nothing specific from rabbitmq but amqp, which points make you mind that the code knows the underlying mq ?

File: src/papi-client.ts [hunk 2]
```diff
@@ -10,6 +10,20 @@ export interface PapiCredentials {
     login: string;
     password: string;
 }
+
+export interface BatchResult {
+    message: string;
+    errors: BatchError[];
+}
+
+export interface BatchError {
```
Comment (thomas-hilaire) [inline src/papi-client.ts]: yes, I was lazy about that

File: src/papi-client.ts [hunk 3]
```diff
@@ -39,18 +53,21 @@ export class PapiClient {
...
+    public waitForBatchSuccess(delay?: number): Promise<BatchResult> {
         this.assertBatchHasBeenStarted();
 
-        let deferred = Promise.defer<string>();
+        let deferred = Promise.defer<BatchResult>();
         let callback = (err, res) => {
             if (err) {
                 deferred.reject(err);
             } else if (res.body.status === "ERROR") {
                 deferred.reject("ERROR: " + res.body.operationDone + "/" + res.body.operationCount);
```
Comment (thomas-hilaire) [inline src/papi-client.ts]: The "reject" does not have to returns the same type than the "success". Rejecting with a string or an exception is a common practice I think.

File: src/importer.ts [hunk 4]
```diff
@@ -30,35 +30,45 @@ export class Importer {
...
+        // We are not able to know which items are in error so we requeue all.
+        // As the import operations are idempotent, process again well imported item won't have any effect.
+        if (batchResult.errors.length === 0) {
+            amqpMessages.forEach(e => e.ack());
+        } else {
+            amqpMessages.forEach(e => amqpChannel.sendToQueue(this.config.amqp.queues.event, e.content, e.properties));
```
Comment (fvn-linagora) [inline src/importer.ts]: This error handling statement could be moved out and passed as a 'publish' function built during MQ connection phase IMO.

Comment (thomas-hilaire) [inline src/importer.ts]: Same thing, I do such thing in the next review, so I don't want to refactor in these commits.

File: src/papi-client.ts (hunk 2 above)
Comment (thomas-hilaire) [inline src/papi-client.ts]: done

File: src/papi-client.ts [hunk 5]
```diff
@@ -60,9 +64,9 @@ export class PapiClient {
         let callback = (err, res) => {
             if (err) {
                 deferred.reject(err);
-            } else if (res.body.status === "ERROR") {
+            } else if (res.body.status === BatchOperationStatus[BatchOperationStatus.ERROR]) {
```
Comment (fvn-linagora) [inline src/papi-client.ts]: I'd suggest this for better readability :
```
} else if (BatchOperationStatus[res.body.status] === BatchOperationStatus.ERROR) {
```

File: src/papi-client.ts [hunk 6]
```diff
@@ -60,9 +64,9 @@ export class PapiClient {
...
                 deferred.reject("ERROR: " + res.body.operationDone + "/" + res.body.operationCount);
-            } else if (res.body.status === "SUCCESS") {
+            } else if (res.body.status === BatchOperationStatus[BatchOperationStatus.SUCCESS]) {
```
Comment (fvn-linagora) [inline src/papi-client.ts]: Same here

File: src/papi-client.ts [hunk 7]
```diff
@@ -92,7 +96,7 @@ export class PapiClient {
     }
 
     private findBatchErrors(batch): BatchError[] {
-        return batch.operations.filter(o => o.status === "ERROR" || o.error);
+        return batch.operations.filter(o => o.error || o.status === BatchOperationStatus[BatchOperationStatus.ERROR]);
```
Comment (fvn-linagora) [inline src/papi-client.ts]: yet again (ie convert o.status to enum before comparing)

File: src/papi-client.ts (hunk 5 above)
Comment (thomas-hilaire) [inline src/papi-client.ts]: It seems a good idea but `BatchOperationStatus[res.body.status]` is resolved as a string by typescript.
`error TS2365: Operator '===' cannot be applied to types 'string' and 'BatchOperationStatus'`

**Pull request merged**